client/attest.go

```diff
@@ -24,6 +33,78 @@ type AttestOpts struct {
 	// firmware event log, where PCRs 0-9 and 14 are often measured. If the two
 	// logs overlap, server-side verification using this library may fail.
 	CanonicalEventLog []byte
+	// Indicates whether the AK certificate chain should be retrieved for validation.
+	// If true, Key.Attest() will construct the certificate chain by making GET requests to
+	// the contents of Key.cert.IssuingCertificateURL.
+	FetchCertChain bool
+}
+
+// Given a certificate, iterates through its IssuingCertificateURLs and returns the certificate
+// that signed it. Returns an error if a valid signing certificate is not found.
+func fetchIssuingCertificate(cert *x509.Certificate) (*x509.Certificate, error) {
+	if cert == nil {
+		return nil, fmt.Errorf("provided certificate is nil")
+	}
+	// For each URL, fetch and parse the certificate, then verify whether it signed cert.
+	// If successful, return the parsed certificate. If any step in this process fails, try the next url.
+	for _, url := range cert.IssuingCertificateURL {
+		resp, err := http.Get(url)
+		if err != nil {
+			log.Printf("failed to retrieve certificate at %v: %v\n", url, err)
+			continue
+		}
+		defer resp.Body.Close()
+
+		if resp.StatusCode != http.StatusOK {
+			log.Printf("certificate retrieval from %s returned non-OK status: %v\n", url, resp.StatusCode)
+			continue
+		}
+		certBytes, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			log.Printf("failed to read response body from %s: %v\n", url, err)
+			continue
+		}
+
+		parsedCert, err := x509.ParseCertificate(certBytes)
+		if err != nil {
+			log.Printf("failed to parse response from %s into a certificate: %v\n", url, err)
+			continue
+		}
+
+		// Check if the parsed certificate signed the current one.
+		err = cert.CheckSignatureFrom(parsedCert)
+		if err == nil {
+			return parsedCert, nil
+		}
+	}
+
+	return nil, fmt.Errorf("did not find valid signing certificate")
+}
+
+// Constructs the certificate chain for the key's certificate, using the provided HTTP client.
+func (k *Key) getCertificateChain() ([][]byte, error) {
+	if len(k.cert.IssuingCertificateURL) > maxIssuingCertificateURLs {
+		return nil, fmt.Errorf("key cert contains too many issuing URLS: got %v, expect no more than %v", len(k.cert.IssuingCertificateURL), maxIssuingCertificateURLs)
+	}
+
+	var certs [][]byte
+	currentCert := k.cert
+	for i := 0; i < maxCertChainDepth; i++ {
+		issuingCert, err := fetchIssuingCertificate(currentCert)
+		if err != nil {
+			return nil, fmt.Errorf("error retrieving certificate chain: %w", err)
```

Comment:
Per discussion over chat, removed error returns from `fetchIssuingCertificate` and `getCertificateChain`. If an error is encountered, `fetchIssuingCertificate` will return nil and `getCertificateChain` will return whatever part of the chain has been constructed so far.

We are avoiding returning errors here because it is possible to have issues retrieving intermediate certificates, but still have enough to allow `VerifyAttestation` to succeed later on. Therefore we do not want `key.Attest` to fail due to cert chain problems.